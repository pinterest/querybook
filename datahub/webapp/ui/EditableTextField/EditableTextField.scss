.EditableTextField {
    .edit-mode-button {
        cursor: pointer;
        opacity: 0;

        display: flex;
        margin-top: 4px;
        margin-right: 4px;
    }

    &:hover {
        .edit-mode-button {
            opacity: 1;
        }
    }

    &.edit-mode .editor-wrapper {
        border: 1px dashed var(--focus-border-color);
        border-radius: var(--border-radius);
    }

    .editor-wrapper {
        display: flex;
        padding: 8px;
        word-break: break-word;
        align-items: flex-start;

        .RichTextEditor {
            flex: 1;
        }

        border: 1px solid transparent;
    }

    .edit-mode-footer {
        display: flex;
        justify-content: flex-end;
        margin-top: 8px;
    }
}
