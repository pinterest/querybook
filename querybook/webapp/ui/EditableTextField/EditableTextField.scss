.EditableTextField {
    .edit-mode-button {
        cursor: pointer;
        opacity: 0.5;

        display: flex;
        margin-top: -2px;
        margin-right: 4px;
    }

    .editor-wrapper {
        display: flex;
        padding: var(--padding);
        word-break: break-word;
        align-items: flex-start;
        border-radius: var(--border-radius-sm);

        .RichTextEditor {
            flex: 1;
        }

        border: 1px solid transparent;
    }

    &:hover {
        .edit-mode-button {
            opacity: 1;
        }
    }

    &.edit-mode .editor-wrapper {
        background-color: var(--bg-light);
    }

    .edit-mode-footer {
        display: flex;
        justify-content: flex-end;
        margin-top: 8px;
    }
}

.FormFieldInputSection {
    .EditableTextField {
        background-color: var(--bg-light);
        border-radius: var(--border-radius-sm);
        .editor-wrapper {
            padding: 0px;
        }
        .edit-mode-button {
            margin: 4px 8px;
        }
    }
}
