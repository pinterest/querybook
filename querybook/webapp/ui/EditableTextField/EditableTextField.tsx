import clsx from 'clsx';
import * as DraftJs from 'draft-js';
import React from 'react';

import { AsyncButton } from 'ui/AsyncButton/AsyncButton';
import { Button } from 'ui/Button/Button';

import { RichTextEditor } from 'ui/RichTextEditor/RichTextEditor';
import './EditableTextField.scss';

export interface IEditableTextFieldProps {
    value: DraftJs.ContentState;
    onSave: (content: DraftJs.ContentState) => Promise<any>;
    className?: string;
}

export const EditableTextField: React.FunctionComponent<IEditableTextFieldProps> = ({
    value,
    onSave,
    className,
}) => {
    const [editMode, setEditMode] = React.useState(false);
    const editorRef = React.useRef<RichTextEditor>(null);

    const toggleEditMode = () => setEditMode(!editMode);
    const setEditorContent = (content: DraftJs.ContentState) => {
        if (editorRef.current) {
            editorRef.current.setContent(content);
        }
    };
    const handleCancel = () => {
        setEditorContent(value);
        toggleEditMode();
    };
    const handleSave = async () => {
        if (onSave && editorRef.current) {
            await onSave(editorRef.current.getContent());
        }
        toggleEditMode();
    };

    React.useEffect(() => {
        setEditorContent(value);
    }, [value]);

    const toggleEditModeButton = !editMode ? (
        <span onClick={toggleEditMode} className="edit-mode-button">
            <i className="fa fa-edit" />
        </span>
    ) : null;

    const editor = (
        <div className="editor-wrapper">
            <RichTextEditor
                value={value}
                readOnly={!editMode}
                ref={editorRef}
            />
            {toggleEditModeButton}
        </div>
    );

    const editModeButtons = editMode ? (
        <div className="edit-mode-footer">
            <Button color="cancel" title="Cancel" onClick={handleCancel} />
            <AsyncButton color="confirm" title="Save" onClick={handleSave} />
        </div>
    ) : null;

    return (
        <div
            className={clsx({
                EditableTextField: true,
                'edit-mode': editMode,
                [className]: Boolean(className),
            })}
        >
            {editor}
            {editModeButtons}
        </div>
    );
};
