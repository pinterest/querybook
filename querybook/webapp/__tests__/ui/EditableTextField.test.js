import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import { EditableTextField } from '../../ui/EditableTextField/EditableTextField';

it('renders without crashing', () => {
    shallow(<EditableTextField value="test" onSave={() => null} />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(
            <EditableTextField value="test" onSave={() => null} />
        );
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});
