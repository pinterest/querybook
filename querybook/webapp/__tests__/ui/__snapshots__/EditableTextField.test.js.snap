// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<div
  className="EditableTextField"
>
  <div
    className="editor-wrapper"
  >
    <RichTextEditor
      readOnly={true}
      value="test"
    />
    <span
      className="edit-mode-button"
      onClick={[Function]}
    >
      <i
        className="fa fa-edit"
      />
    </span>
  </div>
</div>
`;
